File: Problem2.py
'''

Time Complexity : O(n)
Space Complexity : O(n)
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No issues faced

Your code here along with comments explaining your approach

Very similar to 2-sum, we check if a pair of elements exists whose sum is equal to the target and add it to a hashset
The hashset holds all the unique pair of numbers

'''

class Solution:
    def findPairs(self, nums: List[int], k: int) -> int:
        maps = set()
        res = set()

        for i in range(len(nums)):
            if (nums[i]-k) in maps:
                if (nums[i]-k,nums[i]) in res or (nums[i],nums[i]-k) in res:
                    continue
                res.add((nums[i]-k,nums[i]))

            if (nums[i]+k) in maps:
                if (nums[i]+k,nums[i]) in res or (nums[i],nums[i]+k) in res:
                    continue
                res.add((nums[i]+k,nums[i]))

            maps.add(nums[i])

        return len(res)

File: Problem2.js
//https://leetcode.com/problems/k-diff-pairs-in-an-array/submissions/
// Accepted in LeetCode
//tC is O(n) and sc is O(n)


/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var findPairs = function(nums, k) {
    
    let set = new Set(nums);

    let res=0;
    for(let num of set){
        let target = num+k;
        if(set.has(target))
        res++
    }
 // edge case when the target is zero, check for duplicates
    if(k==0){
        let res=0;
        count ={};

        for(let num of nums){
            count[num] = count[num] || 0;
            count[num]++;

            if(count[num] == 2)
            res++
        }
        return res;
    }

    return res;

};